### PR TITLE
Allure testops scenario

### PR DESCRIPTION
**Pull request by andrey-brovko**

### CODE REVIEW

**Review by ILastovka — APPROVED** [review]

File: GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs [hunk 1]
```diff
@@ -79,35 +82,56 @@ public void Sync(List<IFeatureFile> featureFiles)
...
-
+            //TODO:
+            DeleteNotExistingScenarios(allureTestCases, featureFilesTagIds);
             Log.Debug(@$"Synchronization with Allure TestOps finished in: {stopwatch.Elapsed:mm\:ss\.fff}");
         }
 
-        private TestCase CreateTestCase(CreateTestCaseRequest caseRequest)
+        private void DeleteNotExistingScenarios(List<TestCaseContent> allureTestCases, List<ulong> featureFilesTagIds)
+        {
+            //TODO:
```
Comment (vladimir-s-quantori) [inline GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs:103]: It seems to me that it's better to throw `NotImplementedException` here with appropriate message

File: GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs [hunk 2]
```diff
@@ -45,19 +47,125 @@ public CaseContentBuilder(AllureClientWrapper allureClientWrapper, Context conte
...
+            var allureStep = new Quantori.AllureTestOpsClient.Model.Step
+            {
+                Keyword = step.Keyword.Trim(),
+                Name = step.Text
+            };
+
+            allureSteps.Add(allureStep);
+        }
+
+        return allureSteps;
```
Comment (vladimir-s-quantori) [inline GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs:152]: Here you can easily use Linq Select method and return result in one command
And to make it more convenient you can add constructor to `Quantori.AllureTestOpsClient.Model.Step` for keyword and name

File: GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs [hunk 3]
```diff
@@ -62,42 +62,187 @@ public TestCaseOverview GetTestCaseOverview(ulong id)
...
+            }
+
+            Log.Info(updated
+                ? $"Updated: [{currentCase.Id}] {caseToUpdate.CreateTestCaseRequest.Name}"
+                : $"Up-to-date: [{currentCase.Id}] {currentCase.Name}");
+        }
+
+        private bool IsScenarioEqual(TestCaseOverview testCaseOverview, CreateTestCaseRequestExtended caseToUpdate)
+        {
+            if (testCaseOverview.Scenario is null && caseToUpdate.CreateTestCaseRequest.Scenario is null)
```
Comment (vladimir-s-quantori) [inline GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs:109]: null == null? :)

File: GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs [hunk 4]
```diff
@@ -62,42 +62,187 @@ public TestCaseOverview GetTestCaseOverview(ulong id)
...
+            {
+                return true;
+            }
+
+            if (testCaseOverview.Scenario is null && caseToUpdate.CreateTestCaseRequest.Scenario.Steps is not null)
+            {
+                if (caseToUpdate.StepsAttachments.Any())
+                {
+                    UpdateTestCaseStepAttachments(caseToUpdate, testCaseOverview);
```
Comment (vladimir-s-quantori) [inline GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs:118]: Not quite understand why equality comparer updates something

File: GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs [hunk 5]
```diff
@@ -79,35 +82,56 @@ public void Sync(List<IFeatureFile> featureFiles)
...
             }
-
+            //TODO:
+            DeleteNotExistingScenarios(allureTestCases, featureFilesTagIds);
             Log.Debug(@$"Synchronization with Allure TestOps finished in: {stopwatch.Elapsed:mm\:ss\.fff}");
         }
 
-        private TestCase CreateTestCase(CreateTestCaseRequest caseRequest)
+        private void DeleteNotExistingScenarios(List<TestCaseContent> allureTestCases, List<ulong> featureFilesTagIds)
+        {
```
[the remark this answers is not in the document]
Comment (andrey-brovko) [reply GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs:103]: Done

File: GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs (hunk 1 above)
Comment (andrey-brovko) [reply GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs]: Done

File: GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs (hunk 3 above)
Comment (andrey-brovko) [reply GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs:109]: Not exactly. If both Allure and Feature file scenarios are empty we consider this as equality.

File: GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs (hunk 4 above)
Comment (andrey-brovko) [reply GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs]: Good point! Fixed.

File: GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs [hunk 6]
```diff
@@ -45,19 +47,125 @@ public CaseContentBuilder(AllureClientWrapper allureClientWrapper, Context conte
...
+            var backgroundSteps = ExtractSteps(background.Steps.ToList());
+            return backgroundSteps.Concat(scenarioSteps).ToList();
+        }
+
+        return scenarioSteps;
+    }
+
+    private List<Quantori.AllureTestOpsClient.Model.Step> ExtractSteps(List<Step> steps)
+    {
+        var allureSteps = new List<Quantori.AllureTestOpsClient.Model.Step>();
```
Comment (vladimir-s-quantori) [inline GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs:140]: You can implement this to make code more readable:
`using AllureTestCaseStep = Quantori.AllureTestOpsClient.Model.Step` in the usings block

File: GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs [hunk 7]
```diff
@@ -62,42 +62,179 @@ public TestCaseOverview GetTestCaseOverview(ulong id)
...
+                ValidateResponse(response);
+                updated = true;
             }
+
+            Log.Info(updated
+                ? $"Updated: [{currentCase.Id}] {caseToUpdate.CreateTestCaseRequest.Name}"
+                : $"Up-to-date: [{currentCase.Id}] {currentCase.Name}");
+        }
+
+        private bool IsScenarioEqual(TestCaseOverview testCaseOverview, CreateTestCaseRequestExtended caseToUpdate)
```
Comment (vladimir-s-quantori) [inline GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs:114]: I think it should be called `AreTestCasesEqual`

File: GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs [hunk 8]
```diff
@@ -62,42 +62,179 @@ public TestCaseOverview GetTestCaseOverview(ulong id)
...
+                foreach (var id in attachment.Select(step => step.Attachments.FirstOrDefault()!.Id))
+                {
+                    RemoveTestCaseAttachment(id);
+                }    
+            }
+            
+            AddTestCaseStepAttachments(caseToUpdate, testCaseOverview.Id);
+        }
+
+        private static bool IsTestCaseContentEqual(TestCaseOverview currentCase, CreateTestCaseRequestExtended caseToUpdate)
```
Comment (vladimir-s-quantori) [inline GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs:233]: Same: `AreTestCasesContentsEqual`

File: GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs (hunk 2 above)
Comment (andrey-brovko) [reply GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs]: Done

File: GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs (hunk 6 above)
Comment (andrey-brovko) [reply GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs]: Done

File: GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs (hunk 7 above)
Comment (andrey-brovko) [reply GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs]: Fixed

File: GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs (hunk 8 above)
Comment (andrey-brovko) [reply GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClientWrapper.cs]: Fixed

**Review by vladimir-s-quantori — APPROVED** [review]

**Pull request merged**